IPython/core/completer.py

```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    class _CompletionContextType(enum.Enum):
+        ATTRIBUTE = "attribute"  # For attribute completion
+        GLOBAL = "global"  # For global completion
+
+    def _determine_completion_context(self, line):
+        """
+        Determine whether the cursor is in an attribute or global completion context.
+        """
+        # Cursor in string/comment → GLOBAL.
+        if self._is_in_string_or_comment(line):
+            return self._CompletionContextType.GLOBAL
+
+        # Match 'anything-dot-word' at end for attribute context.
+        # Ex: 'obj.', 'np.random.ran'.
+        chain_match = re.search(r"([a-zA-Z_].*)\.([a-zA-Z_][a-zA-Z0-9_]*)?$", line)
+        if chain_match:
+            return self._CompletionContextType.ATTRIBUTE
+
+        # No attribute pattern → GLOBAL.
+        # Ex: 'var', '', '3.', '-42.5.'
+        return self._CompletionContextType.GLOBAL
+
+    def _is_in_string_or_comment(self, text):
+        in_single_quote = False
+        in_double_quote = False
+        in_triple_single = False
+        in_triple_double = False
+        in_template_string = False  # Covers both f-strings and t-strings
+        in_expression = False  # For expressions in f/t-strings
+        i = 0
+
+        while i < len(text):
+            # Check for f-string or t-string start
+            if (
+                i + 1 < len(text)
+                and text[i] in ("f", "t")
+                and (text[i + 1] == '"' or text[i + 1] == "'")
+            ):
+                in_template_string = True
```

Comment:
This is never reset to `False`, here is how it fails - can you add a test case?

![image](https://github.com/user-attachments/assets/bdf42260-73f3-411b-95c0-d569f1976475)